src/gateway_gatt_char_wifi_ssid.erl

```diff
@@ -51,53 +52,96 @@ read_value(State=#state{value=Value}, _) ->
     {ok, Value, State}.
 
 
-handle_signal(SignalID, _Msg, State=#state{wifi_signal=SignalID}) ->
-    %% WiFi overall state has changed
-    {noreply, maybe_notify_value(update_value(State))};
+handle_signal(SignalID, Msg, State=#state{wifi_signal=SignalID}) ->
+    %% WiFi interface state has changed
+    Args = ebus_message:args(Msg),
+    lager:info("wifi_signal: ~p", [Args]),
+    case Args of
+        {ok, ["Connected", true]} ->
+            {noreply, maybe_notify_value(update_value(State, "online"))};
+        {ok, ["Connected", false]} ->
+            {noreply, maybe_notify_value(update_value(State, "disconnect"))};
+        _ ->
+            {noreply, State}
+    end;
 handle_signal(SignalID, Msg, State=#state{service_signal=SignalID}) ->
     %% Signal for the current service. Deal only with changes to the
     %% State attribute
-    case ebus_message:args(Msg) of
-        {ok, ["State", _]} ->
-            {noreply, maybe_notify_value(update_value(State))};
+    Args = ebus_message:args(Msg),
+    lager:info("service_signal: ~p", [Args]),
+    case Args of
+        {ok, ["State", ServiceState]} ->
+            {noreply, maybe_notify_value(update_value(State, ServiceState))};
+        _ ->
+            {noreply, State}
+    end;
+handle_signal(_SignalID, Msg, State=#state{}) ->
+    %% State#state.service_signal /= _SignalID.
+    %% Signal from a different service
+    Args = ebus_message:args(Msg),
+    lager:warning("unexpected service_signal: ~p", [Args]),
+    case Args of
+        {ok, ["State", "online"]} ->
+            NextState = maybe_notify_value(update_value(State, "disconnect")),
+            {noreply, maybe_notify_value(update_value(NextState, "online"))};
         _ ->
             {noreply, State}
     end.
 
-
 %%
 %% Internal
 %%
 
--spec update_value(#state{}) -> #state{}.
-update_value(State=#state{service_path=CurrentPath}) ->
+update_value(State=#state{service_path=CurrentPath}, ServiceState) ->
     %% Start a scan to repopulate list of visible wifi services
     connman:scan(wifi),
-    case {CurrentPath, gateway_config:wifi_services_online()} of
+    case ServiceState of
+        "online" ->
+            on_online_state(CurrentPath, State);
+        _ ->
+            on_offline_states(CurrentPath, State)
+    end.
+
+on_online_state(CurrentPath, State) ->
+    WiFiServicesOnline = gateway_config:wifi_services_online(),
+    case {CurrentPath, WiFiServicesOnline} of
         {CurrentPath, [{_, CurrentPath} | _]} ->
             %% No change in path
             lager:info("WiFi unchanged for ~p at: ~p", [State#state.value, CurrentPath]),
             State;
-        {undefined, []} ->
```

Comment:
Were these wrong? Or just not ever getting used?

Reply:
The `{undefined, []}` and `{CurrentPath, []}` cases are not wrong that processing has just been relocated to `on_offline_states`.